File: Python3/1 - 200/142 Linked List Cycle II.py
# 48ms, 48.48%; 18.3MB, 5.28%
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head: return None
        p1, p2 = head, head
        while True:
            p1 = p1.next
            if not p1: return None
            p1 = p1.next
            if not p1: return None
            p2 = p2.next
            if p1 is p2: break

        cnt = 0
        while True:
            cnt += 1
            p1 = p1.next.next
            p2 = p2.next
            if p1 == p2: break

        p1, p2 = head, head
        for _ in range(1, cnt):
            p2 = p2.next
        while p1 is not p2.next:
            p1 = p1.next
            p2 = p2.next

        return p1

File: 1 - 200/142 Linked List Cycle II.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

// space complexity S(n) = k, 8ms, 100.00%; 9.9MB, 25.55%
class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        if (!head) return nullptr;

        ListNode *p1 = head, *p2 = head;
        while (true) {
            p1 = p1->next;
            if (!p1) return nullptr;
            p1 = p1->next;
            if (!p1) return nullptr;
            p2 = p2->next;
            if (p1 == p2) break;
        }

        int cnt = 0; // length of loop
        while (true) {
            cnt += 1;
            p1 = p1->next->next;
            p2 = p2->next;
            if (p1 == p2) break;
        }

        p1 = p2 = head;
        for (int i = 1; i < cnt; ++i) p2 = p2->next;
        while (p1 != p2->next) p1 = p1->next, p2 = p2->next;
        return p1;
    }
};
